Sources/NIOCertificateHotReloading/NIOCertificateHotReloading.swift

```diff
@@ -0,0 +1,319 @@
+        public var format: Encoding
+
+        /// Initialize a new ``TimedCertificateReloader/PrivateKeyDescription``.
+        /// - Parameters:
+        ///   - location: A ``TimedCertificateReloader/Location``.
+        ///   - format: A ``TimedCertificateReloader/Encoding``.
+        public init(location: Location, format: Encoding) {
+            self.location = location
+            self.format = format
+        }
+    }
+
+    private struct CertificateKeyPair {
+        var certificate: NIOSSLCertificateSource
+        var privateKey: NIOSSLPrivateKeySource
+    }
+
+    private let refreshInterval: TimeAmount
+    private let certificateDescription: CertificateDescription
+    private let privateKeyDescription: PrivateKeyDescription
+    private let state: NIOLockedValueBox<CertificateKeyPair?>
+
+    /// A `NIOSSLContextConfigurationOverride` that will be used as part of the NIO application's TLS configuration.
+    /// Its certificate and private key will be kept up-to-date via the reload mechanism the ``TimedCertificateReloader``
+    /// implementation provides.
+    public var sslContextConfigurationOverride: NIOSSLContextConfigurationOverride {
+        get {
+            var override = NIOSSLContextConfigurationOverride()
+            guard let certificateKeyPair = self.state.withLockedValue({ $0 }) else {
+                return override
+            }
+            override.certificateChain = [certificateKeyPair.certificate]
+            override.privateKey = certificateKeyPair.privateKey
+            return override
+        }
+    }
+
+    /// Initialize a new ``TimedCertificateReloader``.
+    /// - Parameters:
+    ///   - refreshInterval: The interval at which attempts to update the certificate and private key should be made.
+    ///   - certificateDescription: A ``CertificateDescription``.
+    ///   - privateKeyDescription: A ``PrivateKeyDescription``.
+    public init(
+        refreshingEvery refreshInterval: TimeAmount,
+        certificateDescription: CertificateDescription,
+        privateKeyDescription: PrivateKeyDescription
+    ) {
+        self.refreshInterval = refreshInterval
+        self.certificateDescription = certificateDescription
+        self.privateKeyDescription = privateKeyDescription
+        self.state = NIOLockedValueBox(nil)
+
+        // Immediately try to load the configured cert and key to avoid having to wait for the first
+        // reload loop to run.
+        self.reloadPair()
+    }
+
+    /// Initialize a new ``TimedCertificateReloader``.
+    /// - Parameters:
+    ///   - refreshInterval: The interval at which attempts to update the certificate and private key should be made.
+    ///   - certificateDescription: A ``CertificateDescription``.
+    ///   - privateKeyDescription: A ``PrivateKeyDescription``.
+    @available(macOS 13, iOS 16, tvOS 16, watchOS 9, *)
+    public init(
+        refreshingEvery refreshInterval: Duration,
```

Comment:
Do we run the risk of having ambiguity issues with the `Duration`/`TimeAmount` overload? Both have similar static funcs for e.g. `.seconds` etc.

Reply:
I don't actually know the answer to this but at least on the tests (even when adding an availability annotation to match the `Duration` overload) it resolves to using `TimeAmount`. I am unsure if some other combination of factors may cause the compiler to become confused.